### PR TITLE
Add RAdam and LookAhead optimizers

### PR DESCRIPTION
**Pull request by CyberZHG**
Related to #422:

* The name of the optimizer is `RectifiedAdam` in case of conflicting with other abbreviations. (The name appears in the first line of the second page in the paper)
* Add `Lookahead` optimizer wrapper.
* Unit tests for RAdam are compared with the results of the [official implementation](https://github.com/LiyuanLucasLiu/RAdam).
* Unit tests for Ranger (RAdam + Lookahead) are compared with the results of the [implementation](https://github.com/lessw2020/Ranger-Deep-Learning-Optimizer) from the proposer.

### CODE REVIEW

**Comment by Squadrick** [comment]
I think a more useful use of `LookAhead` would be to make a generic wrapper that can take a `tf.keras.optimizers.Optimizer` and apply lookahead. (Similar to [`tfa.optimizers.MovingAverage`](https://github.com/tensorflow/addons/blob/4022d3ac392a34e540c0f856fdec591a9eadc7d3/tensorflow_addons/optimizers/moving_average.py#L44)). 

So the final API for using `Ranger` would look like this:
```python
radam = tfa.optimizers.RectifiedAdam(lr=1e-3)
ranger = tfa.optimizers.LookAhead(radam, step=4, ratio=0.5)
```
This will allow for other optimizers to be used with `LookAhead`.

**Comment by CyberZHG** [comment]
I've added a standalone Lookahead wrapper.

File: tensorflow_addons/optimizers/rectified_adam.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+# limitations under the License.
+# ==============================================================================
+"""Rectified Adam (RAdam) optimizer."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.python import ops
+from tensorflow.python.ops import (math_ops, state_ops,
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/rectified_adam.py]: Look at my comments regarding private/public API in `lookahead.py`

**Comment by CyberZHG** [comment]
It seems that the `mod` operation is not working properly on GPU. I've changed the implementation to use `floordiv` in `Lookahead`. I don't know if there will be any numerical stability issue with the modification.

File: tensorflow_addons/optimizers/lookahead.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    def __init__(self,
+                 optimizer,
+                 sync_period=6,
+                 slow_step_size=0.5,
+                 name="Lookahead",
+                 **kwargs):
+        r"""Wrap optimizer with the lookahead mechanism.
+
+        Args:
+            optimizer: A Tensor or a floating point value.
```
Comment (WindQAQ) [inline tensorflow_addons/optimizers/lookahead.py]: Probably wrong docstring here.

**Comment by seanpmorgan** [comment]
> It seems that the `mod` operation is not working properly on GPU. I've changed the implementation to use `floordiv` in `Lookahead`. I don't know if there will be any numerical stability issue with the modification.

I'm not sure why TF2 isn't using a soft device placement for this. There is no GPU implementation for the overloaded `__mod__`:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/cwise_op_gpu_mod.cu.cc#L22

But I would have assumed it'd fall back to CPU with dynamic kernels implemented. cc @karmel @martinwicke

File: tensorflow_addons/optimizers/lookahead.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+        Args:
+            optimizer: The original optimizer that will be used to compute
+                and apply the gradients.
+            sync_period: An integer. The synchronization period of lookahead.
+                Enable lookahead mechanism by setting it with a positive value.
+            slow_step_size: A floating point value.
+                The ratio for updating the slow weights.
+            name: Optional name for the operations created when applying
+                gradients. Defaults to "RectifiedAdam".
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/lookahead.py]: Rename `"RectifiedAdam"` -> `"Lookahead"`

File: tensorflow_addons/optimizers/lookahead.py [hunk 4]
```diff
@@ -108,11 +108,13 @@ def _init_op(self, var):
     def _look_ahead_op(self, var):
         var_dtype = var.dtype.base_dtype
         slow_var = self.get_slot(var, 'slow')
-        local_step = tf.cast(self.iterations + 1, var_dtype)
-        sync_period = self._get_hyper('sync_period', local_step.dtype)
+        local_step = tf.cast(self.iterations + 1, tf.dtypes.int64)
+        sync_period = self._get_hyper('sync_period', tf.dtypes.int64)
         slow_step_size = self._get_hyper('slow_step_size', var_dtype)
         step_back = slow_var + slow_step_size * (var - slow_var)
-        sync_cond = tf.equal(local_step % sync_period, 0)
```
Comment (facaiy) [inline tensorflow_addons/optimizers/lookahead.py]: > I would have assumed it'd fall back to CPU with dynamic kernels implemented.

I believe so. Have you ever tried `tf.math.floormod`?

Comment (CyberZHG) [reply tensorflow_addons/optimizers/lookahead.py]: I've tried this one and it's not working.

Comment (CyberZHG) [reply tensorflow_addons/optimizers/lookahead.py]: w/blob/d30ba8443710e94d6a1ab1c71a2c61c437e24790/tensorflow/core/kernels/cwise_op_gpu_floor_mod.cu.cc#L22

Comment (facaiy) [reply tensorflow_addons/optimizers/lookahead.py]: Do you use tf.device to force all ops place on GPU? Can you give a minimal reproducing example? Thanks

Comment (CyberZHG) [reply tensorflow_addons/optimizers/lookahead.py]: The outputs are the same as the results of this [CI](https://source.cloud.google.com/results/invocations/d11ec6bf-ede3-40cf-9c10-3112fe54ba12/targets). It fails the [fit_simple_linear_model](https://github.com/CyberZHG/addons/blob/b4cea8189724f88663637daa1df82d035d97cba4/tensorflow_addons/optimizers/lookahead_test.py#L110-L128) test.

Comment (facaiy) [reply tensorflow_addons/optimizers/lookahead.py]: Thanks for the information, really useful! Could you file an issue for it and we might resolve it later?

**Comment by Squadrick** [comment]
LGTM. @CyberZHG Thank you so much for this contribution, I'm sure the community will find the optimizers very useful.

**Pull request merged**